### PR TITLE
[AzureMonitor] add Log Scopes example to Readme

### PR DESCRIPTION
**Pull request by TimothyMothra**
Related to #44621

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md [hunk 1]
```diff
@@ -293,6 +293,35 @@ The Azure Monitor Distro is a distribution package that facilitates users in sen
...
+```csharp
+builder.Services.Configure<OpenTelemetryLoggerOptions>(options =>
+{
+    options.IncludeScopes = true;
+});
+```
+
+Once enabled, you can define a scope using a dictionary or any enumerable of KeyValuePair<string, object>.
+All logs written within the context of the scope will include the specified information.
+Azure Monitor will add these scope values to the Log's CustomProperties.
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md:312]: please clarify the collision behavior for collision within diff. scopes (as they can be nested), and between scopes and log attributes.

Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md:312]: updated.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md [hunk 2]
```diff
@@ -135,6 +133,40 @@ For more information on the OpenTelemetry project, please review the [OpenTeleme
...
 
+### Log Scopes
+
+Log scopes allow you to add additional properties to the logs generated by your application.
+Although the Azure Monitor Exporter does support scopes, this feature is off by default in OpenTelemetry.
+To leverage log scopes, you must explicitly enable them.
+
+To include the scope with your logs, set `OpenTelemetryLoggerOptions.IncludeScopes` to `true` in your application's configuration:
+```csharp
+var loggerFactory = LoggerFactory.Create(builder =>
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md:144]: Add AzureMonitorExporter to the builder.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md [hunk 3]
```diff
@@ -293,6 +293,37 @@ The Azure Monitor Distro is a distribution package that facilitates users in sen
...
+    new("scopeKey", "scopeValue")
+];
+
+using (logger.BeginScope(scope))
+{
+    logger.LogInformation("Example message.");
+}
+```
+
+Note that if you define nested scopes with the same key or multiple keyvaluepairs with the same key, only the first value will be set.
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md:325]: this is still ambigous. "First value" refers to what? 
From source code, scopes are before log statements, so this means scope gets priority? or viceversa?
what about multiple scopes?

Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md:325]: Thanks Cijo, this was more nuanced than I first understood. I used Copilot to help write this new statement.
Can I please get your feedback on this:

>In scenarios involving multiple scopes or a single scope with multiple key-value pairs, if duplicate keys are present, only the first occurrence of the key-value pair from the outermost scope will be recorded.
However, when the same key is utilized both within a logging scope and directly in the log message template, the value specified in the log message template will take precedence.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md [hunk 4]
```diff
@@ -135,6 +133,42 @@ For more information on the OpenTelemetry project, please review the [OpenTeleme
...
+];
+
+using (logger.BeginScope(scope))
+{
+    logger.LogInformation("Example message.");
+}
+```
+
+In scenarios involving multiple scopes or a single scope with multiple key-value pairs, if duplicate keys are present, only the first occurrence of the key-value pair from the outermost scope will be recorded.
+However, when the same key is utilized both within a logging scope and directly in the log message template, the value specified in the log message template will take precedence.
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md:170]:
```suggestion
However, when the same key is utilized both within a logging scope and directly in the log statement, the value specified in the log message template will take precedence.
```

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md [hunk 5]
```diff
@@ -135,6 +133,42 @@ For more information on the OpenTelemetry project, please review the [OpenTeleme
...
+    new("scopeKey", "scopeValue")
+];
+
+using (logger.BeginScope(scope))
+{
+    logger.LogInformation("Example message.");
+}
+```
+
+In scenarios involving multiple scopes or a single scope with multiple key-value pairs, if duplicate keys are present, only the first occurrence of the key-value pair from the outermost scope will be recorded.
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md:169]: this is better, but it is still missing the priority ordering for the duplicate key within the same scope, as shown below. 
```
List<KeyValuePair<string, object>> scope =
[
    new("scopeKey", "scopeValue1")
    new("scopeKey", "scopeValue2")
];

using (logger.BeginScope(scope))
```

OTel SDK upstream does not de-duplicate, so its is upto Exporters, so I'd suggest to cover this with good unit test coverage as well.

Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md:169]: > it is still missing the priority ordering for the duplicate key within the same scope

I thought I covered that case with this quote: `or a single scope with multiple key-value pairs`
Do you have any suggestions on how to rephrase this to make it easier to understand?



> I'd suggest to cover this with good unit test coverage

I already added the code sample you shared. 
I can add more tests specifically around scopes.

https://github.com/Azure/azure-sdk-for-net/blob/3f1749c5734a9ac0435790201aef312aea528f05/sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/E2ETelemetryItemValidation/LogsTests.cs#L69-L83

Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md:169]: I found several tests around scopes in the LogsHelperTests class.
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/LogsHelperTests.cs

I added one more to check behavior when the same key is present in both the scope and the log template.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md [hunk 6]
```diff
@@ -293,6 +293,39 @@ The Azure Monitor Distro is a distribution package that facilitates users in sen
...
+To include the scope with your logs, set `OpenTelemetryLoggerOptions.IncludeScopes` to `true` in your application's configuration:
+```csharp
+builder.Services.AddOpenTelemetry().UseAzureMonitor();
+builder.Services.Configure<OpenTelemetryLoggerOptions>(options =>
+{
+    options.IncludeScopes = true;
+});
+```
+
+Once enabled, you can define a scope using a dictionary or any enumerable of KeyValuePair<string, object>.
```
Comment (reyang) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md:311]:
```suggestion
Once enabled, you can define a scope using a dictionary or any enumerable of `KeyValuePair<string, object>`.
```

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/LogsHelperTests.cs [hunk 7]
```diff
@@ -415,6 +415,47 @@ public void DuplicateKeysInLogRecordAttributesAndLogScope()
...
+            const string duplicateScopeValue2 = "Another duplicate scope value";
+
+            // Act.
+            using (logger.BeginScope(new List<KeyValuePair<string, object>>
+            {
+                new KeyValuePair<string, object>(expectedScopeKey, expectedScopeValue),
+                new KeyValuePair<string, object>(expectedScopeKey, duplicateScopeValue),
+            }))
+            {
+                logger.LogInformation($"Some log information message. {{{expectedScopeKey}}}.", duplicateScopeValue2);
```
Comment (reyang) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/LogsHelperTests.cs:446]: Why using string interpolation?

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md (hunk 6 above)
Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md:311]: Good catch, let me make this change on my end so I get both Readme's updated. :)

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/E2ETelemetryItemValidation/LogsTests.cs [hunk 8]
```diff
@@ -57,19 +57,36 @@ public void VerifyLog(LogLevel logLevel, string expectedSeverityLevel)
...
+            {
+                new("scopeKey1", "scopeValue3")
+            };
+
+            using (logger.BeginScope(scope1))
+            using (logger.BeginScope(scope2))
+            {
+                logger.Log(
+                    logLevel: logLevel,
+                    eventId: 0,
```
Comment (reyang) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/E2ETelemetryItemValidation/LogsTests.cs:85]: Hmm, I think 0 should never be used.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/LogsHelperTests.cs (hunk 7 above)
Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/LogsHelperTests.cs:446]: This approach was taken just for brevity. It keeps the end-to-end scenario within the method block. All of the tests in this class/project are using the same approach. 

These were initially written before the best practices were updated to recommend structured logging. We've never made it a priority to refactor these tests, and new tests are following the existing pattern.

I'll bring this up in our next standup to discuss if we want to spend the time to refactor this project. I think it would add value because some devs will refer to the examples in the unit tests. But we need to weigh that against covering our other test gaps.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/E2ETelemetryItemValidation/LogsTests.cs (hunk 8 above)
Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/E2ETelemetryItemValidation/LogsTests.cs:85]: I agree, we have a special case in the exporter where we do not honor zero as the event ID.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/LogsHelperTests.cs (hunk 7 above)
Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/LogsHelperTests.cs:446]: We should consider using the AI assistant to upgrade these code.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/E2ETelemetryItemValidation/LogsTests.cs (hunk 8 above)
Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/E2ETelemetryItemValidation/LogsTests.cs:85]: Spoke to Raj offline, I'm going to fully address this in another PR.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md [hunk 9]
```diff
@@ -135,6 +133,42 @@ For more information on the OpenTelemetry project, please review the [OpenTeleme
...
+{
+    builder.AddOpenTelemetry(options =>
+    {
+        options.AddAzureMonitorLogExporter(o => o.ConnectionString = "InstrumentationKey=00000000-0000-0000-0000-000000000000");
+        options.IncludeScopes = true;
+    });
+});
+```
+
+Once enabled, you can define a scope using a dictionary or any enumerable of `KeyValuePair<string, object>`.
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md:154]: When using `ILogger`
  [scopes](https://learn.microsoft.com/dotnet/core/extensions/logging#log-scopes),
  use a `List<KeyValuePair<string, object?>>` or `IReadOnlyList<KeyValue<string,
  object?>>` as the state for best performance. Use an
  `IEnumerable<KeyValue<string, object?>>` when performance is *NOT* a critical
  requirement.

use the above wording. I had it from a previous discussion. It should ideally be in OTel upstream, but OTel has been shying away from Log Scopes... so lets keep it here atleast.

Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md:154]: > Use an IEnumerable<KeyValue<string, object?>> when performance is NOT a critical requirement.

This seems weird to me. I'm thinking we shouldn't include non-recommended approaches.
Also, getting deep into the performance details of ILogger seems off-topic for Azure Monitor. 
That seems more appropriate for ILogger's own docs

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md [hunk 10]
```diff
@@ -293,6 +293,43 @@ The Azure Monitor Distro is a distribution package that facilitates users in sen
 
 Refer to [`Program.cs`](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Demo/Program.cs) for a complete demo.
 
+### Log Scopes
+
+Log [scopes](https://learn.microsoft.com/dotnet/core/extensions/logging#log-scopes) allow you to add additional properties to the logs generated by your application.
+Although the Azure Monitor Exporter does support scopes, this feature is off by default in OpenTelemetry.
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md:299]:
```suggestion
Although the Azure Monitor Distro does support scopes, this feature is off by default in OpenTelemetry.
```

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md [hunk 11]
```diff
@@ -293,6 +293,43 @@ The Azure Monitor Distro is a distribution package that facilitates users in sen
...
+To leverage log scopes, you must explicitly enable them.
+
+To include the scope with your logs, set `OpenTelemetryLoggerOptions.IncludeScopes` to `true` in your application's configuration:
+```csharp
+var loggerFactory = LoggerFactory.Create(builder =>
+{
+    builder.AddOpenTelemetry(options =>
+    {
+        options.AddAzureMonitorLogExporter(o => o.ConnectionString = "InstrumentationKey=00000000-0000-0000-0000-000000000000");
+        options.IncludeScopes = true;
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md:309]: We should not use `builder.AddOpenTelemetry()` here, rather configure API should be used.

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md:309]: https://github.com/Azure/azure-sdk-for-net/issues/43623

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md [hunk 12]
```diff
@@ -135,6 +133,43 @@ For more information on the OpenTelemetry project, please review the [OpenTeleme
...
+Although the Azure Monitor Exporter does support scopes, this feature is off by default in OpenTelemetry.
+To leverage log scopes, you must explicitly enable them.
+
+To include the scope with your logs, set `OpenTelemetryLoggerOptions.IncludeScopes` to `true` in your application's configuration:
+```csharp
+var loggerFactory = LoggerFactory.Create(builder =>
+{
+    builder.AddOpenTelemetry(options =>
+    {
+        options.AddAzureMonitorLogExporter(o => o.ConnectionString = "InstrumentationKey=00000000-0000-0000-0000-000000000000");
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md:148]: Shall we use a new API introduced in beta version?

Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md:156]: Are you referring to the `SDK.CreateLoggerProviderBuilder()`?
This API is still Experimental. I don't think we should include it in our Readme/docs until it ships and we can take it as a stable dependency.
https://github.com/open-telemetry/opentelemetry-dotnet/blob/36c586ddcc62cc1e4da82c6f9f65dba16b78161a/src/OpenTelemetry/Sdk.cs#L81

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md (hunk 11 above)
Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md:317]: Done

**Review by rajkumar-rangaraj — APPROVED** [review]
